students/luke/s05/mailroom.py

```diff
@@ -0,0 +1,143 @@
+                add_donations()
+            elif action == "w":
+                write_thankyous()
+            elif action == "c":
+                create_report()
+            else:
+                done = True
+        except EOFError:
+            done = True
+        except KeyboardInterrupt:
+            done = True
+    return
+
+
+def build_thankyou(name):
+    """ Apply thankyou format string to name """
+    return (THANK_YOU.format(name, sum(donor_list[name])))
+
+
+def list_donors():
+    """ return list donor names """
+    donors = ""
+    for n in donor_list:
+        donors += n + "\n"
+    return donors[:-1]  # Trim trailing newline
+
+
+def add_donation(name, amount):
+    """ add donation for name.  add name if it doesn't exist """
+    if not donor_list.get(name):
+        print("Adding new donor: " + name)
+        donor_list[name] = []
+
+    donor_list[name].append(donation_amount)
+
+
+def add_donations():
+    """ Add donation history for a(n optionally new) user """
+    done = False
+    while not done:
+        try:
+            name = input("Enter donor name (or \"list\" for list): ").lower()
+        except EOFError:
+            return
+        except KeyboardInterrupt:
+            return
+
+        if name == "list":
+            print(list_donors())
+            continue
+
+        moredonations = True
+        while moredonations:
+            try:
+                value = input("Enter donation amount or [enter] finished: ")
+            except EOFError:
+                break
+            except KeyboardInterrupt:
+                break
+            if not donation_amount:
+                break;
+
+            try:
+                donation_amount = int(value)
+            except ValueError:
+                print("Invalid input, reenter.")
+                continue
+
+            add_donation(name, donation_amount)
+
+        done = True
+        print(build_thankyou(name))
+    return
+
+
+def write_thankyous():
+    """ write thankyou strings to files, one per donor """
+    for d in donor_list:
+        with open(OUTFILE.format(d), 'w') as outfile:
+            outfile.write(build_thankyou(d))
+
+
+def generate_stats(name):
+    """ Return total given, number of gifts, average gift for name """
+    total = sum(donor_list[name])
+    number = len(donor_list[name])
+    average = total / number
+    return total, number, average
+
+def create_report():
+    """ Display table: Donor Name | Total Given | Num Gifts | Average Gift """
+    print("Donor Name | Total Given | Num Gifts | Average Gift")
+    for d in donor_list:
+        total, number, average = generate_stats(d)
+        print(f"{d:10} | "
+              f"{total:11.2f} | "
+              f"{number:9d} | "
+              f"{average:12.2f}")
+    return
+
+
+# Non-interactive unit tests
+assert(build_thankyou("dan") == "Thank you, dan, for your donation(s) of $5!\n")
+assert(generate_stats("carol") == (106, 3, 106/3))
+
```

Comment:
Good start to testing. Generate_stats is a great candidate for another test! I like how you kept the logic and the printing separate. Makes testing much easier! :)